src/components/tl/ucp/gather/gather_knomial.c

```diff
@@ -0,0 +1,171 @@
+#include "config.h"
+#include "gather.h"
+#include "core/ucc_progress_queue.h"
+#include "tl_ucp_sendrecv.h"
+#include "utils/ucc_math.h"
+
+#define SAVE_STATE(_phase)                                                     \
+    do {                                                                       \
+        task->gather_kn.phase = _phase;                                        \
+    } while (0)
+
+void ucc_tl_ucp_gather_knomial_progress(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_ucp_task_t *task      = ucc_derived_of(coll_task, ucc_tl_ucp_task_t);
+    ucc_coll_args_t *  args      = &TASK_ARGS(task);
+    ucc_tl_ucp_team_t *team      = TASK_TEAM(task);
+    ucc_rank_t         team_size = UCC_TL_TEAM_SIZE(team);
+    ucc_rank_t         rank      = UCC_TL_TEAM_RANK(team);
+    ucc_rank_t         size      = UCC_TL_TEAM_SIZE(team);
+    ucc_rank_t         root      = (ucc_rank_t)args->root;
+    uint32_t           radix     = task->gather_kn.radix;
+    ucc_rank_t         vrank     = (rank - root + size) % size;
+    ucc_memory_type_t  mtype     = args->src.info.mem_type;
+    size_t             data_size =
+        args->src.info.count * ucc_dt_size(args->src.info.datatype);
```

Comment:
I think it is ok as is, because:
- if rank=root & INPLACE, the src buffer info is copied from the dst buffer info in the start function, see src/components/tl/ucp/gather/gather_knomial.c:149
- otherwise, the src buffer is initialized by the user